### ISSUE TITLE
Provide (non-field-streaming) record mode

### ISSUE THREAD

**Issue by fsteeg** [metafacture/metafacture-fix#35]
Metafacture has a streaming architecture not just for individual records (which makes sense, we don't want to load all records into memory before processing them), but even on the field level. This would allow us to process something like unbounded graphs, where we have no records, just attributes/fields. In practice though, we do have records coming in from our source data systems. 

Transformation rules often depend on the values of other fields in the same record, and the field-level streaming architecture makes these cases complicated, and often requires complex `flushWith` and `reset` settings. So one of the original goals for metafacture-fix is to provide a non-field-streaming record mode (see https://slides.lobid.org/metafacture-2020/#/68).

It's a bit like XML parsing with SAX vs. DOM. Instead of having a single attribute-value-pair at a time, and being required to understand and correctly handle where we are in the event stream, we want to load all data from a particular record into an object, making all fields available for direct manipulation and querying by the user. This will also get us closer to the way Catmandu::Fix works.

As a simple example, let's assume we want to add `type: Organization` if the `name` field contains `University`. Currently, we'd have to intercept the `name` field, and emit it as `type`, using a `constant`, something like:

```
do map(name, type)
    contains(string: University)
    constant(value: Organization)
end
```

In a record mode, we'd be able to express it like this:

```
if contains(name, University)
    add_field(type, Organization)
end
```

The issue quickly becomes more complex, e.g for cases where we want to do something similar, but on multiple values, like mapping all simple-string-authors from the input to objects (with `name` and `type`) in the output, where we have to flush with the record at one point to collect all authors, and flush with each entity to get correct values in each output object.

@blackwinter: Since we discussed this topic offline, does this issue description cover what you are expecting from this? Feel free to add your own use cases, so we get a good understanding of what our goal is here.

**Comment by blackwinter** [comment]
Awesome. I believe that's a great summary. Thanks!

As for our use cases, here are some Morph snippets that we hope can be handled more easily with Fix:

```xml
<combine name="@portfolio" value="${value}" sameEntity="true">
  <if>
    <data source="POR  .[MA]">
      <equals string="49HBZ_FHA" />
    </data>
  </if>
  <data name="value" source="POR  .a" />
</combine>
<entity name="DateFirst">
  <combine name="date" value="${value}" sameEntity="true">
      <if>
        <equalsFilter name="" value="" flushWith="POC  .a" flushIncomplete="false" reset="true">
          <data source="@portfolio" />
          <data source="POC  .a" />
        </equalsFilter>
      </if>
      <data name="value" source="POC  .b" />
    </combine>
</entity>
```

```xml
<entity name="RSWK[]" flushWith="record">
  <entity name="" flushWith="600??|6[15][01]??|648??" sameEntity="true">
    <if>
      <all flushWith="600??|6[15][01]??|648??" sameEntity="true">
        <data source="600??.2|6[15][01]??.2|648??.2">
          <regexp match="^(gnd|swd)$" />
        </data>
      </all>
    </if>
    <data name="subjectTopicName" source="650??.a">
      <occurrence only="1" sameEntity="true" />
      <replace pattern="\.?$" with="" />
    </data>
    <entity name="subjectIdentifier[]" flushWith="600??|6[15][01]??|648??" sameEntity="true">
      <data name="" source="600??.0|6[15][01]??.0|648??.0" />
    </entity>
  </entity>
</entity>
```

```xml
<entity name="TitleSuper[]" flushWith="record">
  <if>
    <none flushWith="record">
      <data source="830??" />
    </none>
  </if>
  <entity name="" flushWith="490[01] " sameEntity="true" >
    <entity name="titleSuper[]" flushWith="490[01] " sameEntity="true">
      <data source="490[01] .a" name="" />
    </entity>
  </entity>
</entity>
<entity name="TitleSuper[]" flushWith="record">
  <entity name="" flushWith="830 ?" sameEntity="true">
    <entity name="titleSuper[]" flushWith="830 ?" sameEntity="true">
      <data source="830 ?.a" name="" />
    </entity>
  </entity>
</entity>
```

```xml
<entity name="PersonContributor[]" flushWith="record">
  <entity name="" flushWith="record">
    <if>
      <all flushWith="record">
        <data source="100[013] .[4e]" />
        <none flushWith="record">
          <data source="100[013] .[4e]">
            <regexp match="^([Aa]ut|[Cc]re)" />
          </data>
        </none>
      </all>
    </if>
    <data name="personName" source="100[013] .a">
  </entity>
  <entity name="" flushWith="700[013] " sameEntity="true">
    <if>
      <all flushWith="700[013] " sameEntity="true">
        <data source="700[013] .[4e]" />
        <none flushWith="700[013] " sameEntity="true">
          <data source="700[013] .[4e]">
            <regexp match="^([Aa]ut|[Cc]re)" />
          </data>
        </none>
      </all>
    </if>
    <data name="personName" source="700[013] .a">
  </entity>
</entity>
```

There were also some cases that we weren't able to solve at all (yet). I'll try to reconstruct them as time permits.

**Comment by blackwinter** [comment]
Alma use case "local fields" (see also hbz/lobid-resources#1226): Output NZ fields (authors, descriptions, ISBNs, etc.) with IZ-local fields overriding those with, say, identical indicators. In Java pseudo-code it might look something like this:

```java
final String memberID = "...";  // IZ code

final Map<String, Field> map = new HashMap<>();
final List<Field> allFields = record.getFieldsByTag("100");

allFields.forEach(field -> {
  if (field.getSubfield("M") == null) {
    // Add NZ field
    map.put(field.getTagWithIndicators(), field);
  }
});

allFields.forEach(field -> {
  if (memberID.equals(field.getSubfield("M"))) {
    // Add/override with IZ-local field
    map.put(field.getTagWithIndicators(), field);
  }
});

final List<Field> filteredFields = map.values();

// Now map filtered fields to output fields...
```

Until proven otherwise, I'll maintain that this is impossible to solve with Metamorph. Whether Metafix can provide the necessary means, remains to be seen; as a last resort, it should be possible to delegate to a custom Java function (see #14).

**Comment by TobiasNx** [comment]
I played around with Catmandu Fix and rebuild the workflow for Duepublico: https://github.com/TobiasNx/Catmandu-Testing/tree/master/DuePub-Test

I think we can take some inspiration from how Catmandu FIX is different to the current status of Metafacture FIX. It also shows the list-bind in action.

I noted some findings in the readme.

**Comment by fsteeg** [comment]
In the initial approach for implementing Fix, which is used in our current OERSI workflows, I tried to reuse much of the stream-based Metamorph classes. While this got us to a certain point quickly, improving on that turned out to get very complex.

A second take at implementing Fix, ignoring all the stream-based Metamorph classes, turned out much cleaner and easier to work with (see commits above, starting with https://github.com/metafacture/metafacture-fix/commit/f328a1360bbb4e576133db9dd1741a47591b234a), so I'm confident moving forward with that.

We discussed our next steps and a broad time frame today, and this is our current plan:

- [ ] Implement the Catmandu Fix workflow from https://github.com/TobiasNx/Catmandu-Testing/tree/master/DuePub-Test in Metafix (@fsteeg & @TobiasNx, until end of 9/2021)
- [ ] Port the current OERSI (fix, morph) workflows to the new Metafix implementation (@TobiasNx & @fsteeg, until end of 12/2021)
- [ ] Evaluate moving our Alma transformation to Metafix (@TobiasNx, @dr0i, @fsteeg, starting 1/2022)

**Comment by TobiasNx** [comment]
https://github.com/LibreCat/Catmandu/blob/12fd9e44a4813c92ab7b5c3f80e9b9b709d9a8e8/lib/Catmandu/Fix.pm

this might also be interesting for you

**Comment by fsteeg** [comment]
Following up on our plan from https://github.com/metafacture/metafacture-fix/issues/35#issuecomment-902701987 as it's now end of 9/2021:

> Implement the Catmandu Fix workflow from https://github.com/TobiasNx/Catmandu-Testing/tree/master/DuePub-Test in Metafix (@fsteeg & @TobiasNx, until end of 9/2021)

It's not yet fully working, but the basic structure of the output data looks good. I've pushed the current work in progress to the `metafix` branch in OERSI: https://gitlab.com/oersi/oersi-etl/-/blob/metafix/data/experimental/duepub-metafacture.fix (run with `./gradlew clean run --args 'data/experimental/duepub-metafacture.flux'`)

@TobiasNx: With this, you can start using the new Metafix, either continuing in that workflow, or for another workflow (see https://gitlab.com/oersi/oersi-etl/-/issues/88#note_687591370).

@blackwinter: I've published `0.2.0-rc3` to GitHub Packages to make the current state of the `35-metafix` branch available. There are still some conceptual problems and basic things that are not working (see disabled tests).

**Comment by TobiasNx** [comment]
What I spot at once is that the bind `do list` is operating on a single field and not on an array:

```
do list("path": "metadata.mods.identifier.value", "var": "identifier")
    if all_match("identifier","^https://duepublico2.uni-due.de/receive/.*")
        copy_field("identifier","id")
    end
end
```

(See: https://gitlab.com/oersi/oersi-etl/-/blob/metafix/data/experimental/duepub-metafacture.fix#L24)


In Catmandu `do list` can only work with an array:
```
do list(path:_metadata.mods.identifier,var:a)
	if all_match(a._body,"^https://duepublico2.uni-due.de/receive/.*")
		copy_field(a._body,id)
	end
end
```

`do list` works trough every element of an array (either objects or strings) and restarts again.

**Comment by blackwinter** [comment]
> I've published 0.2.0-rc3 to GitHub Packages to make the current state of the 35-metafix branch available.

Thanks! I've opened #57 to also allow us to declare a source dependency in limetrans in case we need to make modifications (see below).

> There are still some conceptual problems and basic things that are not working (see disabled tests).

I'm slowly making myself familiar with the current state of development. The first "surprise" was that it doesn't flatten the event stream by default (like Metamorph does), or at all, apparently ("TODO: keep flattener as option?"). It also doesn't emit the record ID as `_id` (again, like Metamorph does). I'm wondering what level of compatibility should be aimed for.

**Comment by TobiasNx** [comment]
> I'm slowly making myself familiar with the current state of development. The first "surprise" was that it doesn't flatten the event stream by default (like Metamorph does), or at all, apparently ("TODO: keep flattener as option?").

Since the nested structure is kept it allows to map parent level fields and just adapt the lower levels.
